### ISSUE TITLE
miRCounts

### ISSUE THREAD

**Issue by drosofff** [ARTbio/tools-artbio#140]
I propose to use [MirParser](https://github.com/ARTbio/tools-artbio/tree/master/tools/MirParser) as the basis to code an "all-in-one" tool to counts reads that align to miRNAs.

Inputs

- [x] A fastq file of **clipped** sequence reads
- [x] ~~an adapter sequence to be clipped~~
- [x] A fasta file of pre-miR (stem-loop sequences) with an associated GFF3 describing the mature miRs
- [x] or a reference genome with an associated GFF3 describing both pre-mirs and mature mirs (to be discussed). Note that miRBase does not seem to evolve since a couple of years, whereas genomes releases include new annotation on miRNAs

Outputs

- [ ] A count table of reads on pre-mirs (for miRNA transcription)
- [ ] A count table of reads on mature mirs as defined by GFF3 annotations
- [ ] optionally a representation of reads as provided in miRBase or a graphical representation
These outputs would be ready for statistical analysis by DESeq or edgeR

Basically, the tool would execute pieces of codes that already exist:

- [x] ~~Clip adapter~~
- [ ] Index bowtie reference (or choose a locally installed index ?)
- [ ] align reads on bowtie reference
- [ ] Count reads using Bam/Sam alignments and pysam
- [ ] Output count table(s) and graphical view of alignments if requested

I create a tool development folder [tools-arbio/tools/miRCounts](https://github.com/ARTbio/tools-artbio/tree/mircounts/tools/mircounts)

**Comment by drosofff** [comment]
I propose to associate more closely miRCounts with miRBase:
Indeed, the choice of the miR sequences, the download of their sequences and associated GFF from miRBase can be done programmatically.

We can evolve the tool latter so that it can works with organisms not referenced in miRBase, or adapt to users that know how to manipulate specific genomes and annotation.

Thus we can start with a first tool where the user is responsible for downloading Hairpin.fa and GFF3 from miRBase. Then we can automate this in a next version. While automating, it will be possible to expose the code so that it can adapt to any reference and GFF

**Comment by drosofff** [comment]
@ccedmendoza I changed my mind since our last discussion yesterday.
Actually I think that mircount should do what it says it does: counting miR reads !
Thus here, I propose to continue to refactor a counting tool working on ONE fastq file as input and outputing premiR and mature miR counts, and optionally the mir coverage for that one fastq file.
Then it is open to the user to use the option multi-input that will generate multiple jobs. and Finally there are tools to join side by side 2-column tabular files if a global view of counts is needed (but this is not required for DESeq or EdgeR wrapper that work on multiple 2-column datasets in history.

**Then** It gaves me the idea that the current MirParser tool can be refactored to focus on comparison side by side of mir coverages in various samples. for instance MirParser could become MirCoverage in the mississippi tool suite.

I hope you follow me. If not, we can debrief any time
